### PR TITLE
refactor collkernel in hetfrz_classnuc.F90

### PR DESCRIPTION
**Pull request by mingxuanwupnnl**
refactor subroutine collkernel in hetfrz_classnuc.F90 
After discussion with Balwinder, we think the do loop within collkernel does not make much sense, as there are many if statements and the loop length is only 3. We then let it call calculate_collkernell_sub three times.

### CODE REVIEW

File: components/eam/src/physics/cam/hetfrz_classnuc.F90 [hunk 1]
```diff
@@ -761,128 +770,236 @@ pure function get_Aimm(vwice, rgimm, temperature, dg0imm) result(Aimm)
...
+                                 latvap, viscos_air, Re, Ktherm_air, Ktherm_bc, Pr, Tdiff, & ! in
+                                 Kcoll_bc)                                                   ! out
+
+
+   call calculate_collkernel_sub(temperature, pressure, rho_air, r3lx, r_dust_a1, lambda, &    ! in
+                                 latvap, viscos_air, Re, Ktherm_air, Ktherm_dust, Pr, Tdiff, & ! in
+                                 Kcoll_dust_a1)                                                ! out
+
+
+   call calculate_collkernel_sub(temperature, pressure, rho_air, r3lx, r_dust_a3, lambda, &    ! in
```
Comment (tangshuaiqi) [inline components/eam/src/physics/cam/hetfrz_classnuc.F90:847]: yes, I think it is much clear than `do i=1,3` loop

File: components/eam/src/physics/cam/hetfrz_classnuc.F90 [hunk 2]
```diff
@@ -761,128 +770,236 @@ pure function get_Aimm(vwice, rgimm, temperature, dg0imm) result(Aimm)
...
+
+   Q_heat = Ktherm_air/r3lx*(1._r8 + 0.3_r8*Re**0.5_r8*Pr**0.33_r8)*Tdiff
+
+   K_thermo_cotton = 4._r8*pi*r3lx*r3lx*f_t*Q_heat/pressure
+
+   K_diffusio_cotton = -(1._r8/f_t)*(rh2o*temperature/latvap)*K_thermo_cotton
+   K_total = 1.e6_r8*(K_brownian + K_thermo_cotton + K_diffusio_cotton)  ! convert m3/s -> cm3/s
+
+   ! set K to 0 if negative
+   if (K_total .lt. 0._r8) K_total = 0._r8
```
Comment (tangshuaiqi) [inline components/eam/src/physics/cam/hetfrz_classnuc.F90]: change .lt. to <

**Review by tangshuaiqi — APPROVED** [review]
It looks very clear. I only have one minor catch.

**Review by brigaudet — APPROVED** [review]
Looks good to me.

**Review by jeff-cohere — APPROVED** [review]

File: components/eam/src/physics/cam/hetfrz_classnuc.F90 [hunk 3]
```diff
@@ -761,128 +770,236 @@ pure function get_Aimm(vwice, rgimm, temperature, dg0imm) result(Aimm)
...
+   implicit none
+   real(r8), intent(in) :: tc  ! temperature [deg C]
+
+   real(r8) :: viscos_air      ! dynamic viscosity of air [kg/m/s]
+   real(r8), parameter :: coeff_a = -1.2e-5_r8
+   real(r8), parameter :: coeff_b = 0.0049_r8
+   real(r8), parameter :: coeff_c = 1.718_r8
+
+   viscos_air = (coeff_c + coeff_b*tc + coeff_a*tc*tc)*1.e-5_r8
+
```
Comment (mingxuanwupnnl) [inline components/eam/src/physics/cam/hetfrz_classnuc.F90:931]: NBFB issue caused by this line, fixed by changing calculation sequence and make tc**2 to * tc * tc

**Comment by mingxuanwupnnl** [comment]
all tests passed
Screenshot 2023-03-01 at 11 22 45 PM

**Pull request merged**